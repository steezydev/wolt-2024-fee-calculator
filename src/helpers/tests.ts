const localStorageMock = (function () {
    let store: Record<string, string> = {};
    return {
        getItem(key: string) {
            return store[key] || null;
        },
        setItem(key: string, value: string) {
            store[key] = value.toString();
        },
        clear() {
            store = {};
        },
    };
})();

const matchMediaMock = jest.fn().mockImplementation((query) => ({
    matches: query === '(prefers-color-scheme: dark)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
}));

export const mockContext = () => {
    beforeAll(() => {
        Object.defineProperty(window, 'localStorage', {
            value: {
                getItem: jest.fn().mockImplementation((key: string) => {
                    return localStorageMock.getItem(key);
                }),
                setItem: jest
                    .fn()
                    .mockImplementation((key: string, value: string) => {
                        localStorageMock.setItem(key, value);
                    }),
                clear: localStorageMock.clear,
            },
            writable: true,
        });

        window.matchMedia = matchMediaMock;
    });

    beforeEach(() => {
        (window.localStorage.getItem as jest.Mock).mockClear();
        (window.localStorage.setItem as jest.Mock).mockClear();
        matchMediaMock.mockClear();
    });
};

export const setupFakeDate = () => {
    beforeEach(() => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date('2024-01-01'));
    });

    afterEach(() => {
        jest.useRealTimers();
    });
};

export const mockScrollTo = () => {
    beforeEach(() => {
        Element.prototype.scrollTo = jest.fn();
    });

    afterEach(() => {
        (Element.prototype.scrollTo as jest.Mock).mockRestore();
    });
};
