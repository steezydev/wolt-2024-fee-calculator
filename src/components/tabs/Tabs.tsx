import ButtonUnstyled from '@/components/button/ButtonUnstyled';
import { classNames } from '@/helpers/classNames';
import { TabsPanelProps } from '@/types/props/TabsPanelProps';
import { TabsProps } from '@/types/props/TabsProps';
import { TabsTabProps } from '@/types/props/TabsTabProps';
import React, { ReactElement, useMemo, useState } from 'react';

import Panel from './TabsPanel';
import Tab from './TabsTab';

const Tabs = ({ children, className }: TabsProps) => {
  const tabsAndPanels = useMemo(() => {
    const tabs: ReactElement<TabsTabProps>[] = [];
    const panels: ReactElement<TabsPanelProps>[] = [];
    React.Children.forEach(children, (child) => {
      if (React.isValidElement(child)) {
        if (child.type === Tab) {
          tabs.push(child as ReactElement<TabsTabProps>);
        } else if (child.type === Panel) {
          panels.push(child as ReactElement<TabsPanelProps>);
        }
      }
    });
    return { tabs, panels };
  }, [children]);

  const [activeTab, setActiveTab] = useState<string>(
    tabsAndPanels.tabs[0]?.props.value || ''
  );

  const handleTabClick = (value: string) => {
    setActiveTab(value);
  };

  return (
    <div className={classNames('flex flex-col', className)}>
      <div className='flex w-full rounded-xl'>
        {tabsAndPanels.tabs.map((tab, index) => (
          <ButtonUnstyled
            id={tab.props.id}
            ariaLabel={tab.props.ariaLabel}
            key={tab.props.id || index}
            className={classNames(
              'py-1 px-4 w-full',
              index === 0 ? 'rounded-l-xl' : '',
              index === tabsAndPanels.tabs.length - 1 ? 'rounded-r-xl' : '',
              activeTab === tab.props.value
                ? 'bg-primary-300 text-white'
                : 'bg-primary-100 hover:bg-primary-300/80 hover:text-white text-black dark:text-white',
              tab.props.className
            )}
            onClick={() => handleTabClick(tab.props.value)}
          >
            {tab.props.children}
          </ButtonUnstyled>
        ))}
      </div>
      <div>
        {tabsAndPanels.panels.map((panel, index) => (
          <div
            className={classNames(
              'h-full',
              activeTab === panel.props.value ? 'block' : 'hidden',
              panel.props.className
            )}
            key={panel.props.value || index}
          >
            {panel.props.children}
          </div>
        ))}
      </div>
    </div>
  );
};

Tabs.Panel = Panel;
Tabs.Tab = Tab;

export default Tabs;
