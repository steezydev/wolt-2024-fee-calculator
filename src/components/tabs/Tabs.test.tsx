import Tabs from './Tabs';
import { tabsEventTests, tabsRenderTests } from './Tabs.tests';

tabsRenderTests('Tabs', Tabs);
tabsEventTests('Tabs', Tabs);
