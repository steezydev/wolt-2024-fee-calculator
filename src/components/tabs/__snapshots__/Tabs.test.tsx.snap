// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs when asked to render with additional class renders with additional class 1`] = `
<div
  className="flex flex-col some-class"
>
  <div
    className="flex w-full rounded-xl"
  >
    <button
      aria-label="Tab 1 label"
      className="py-1 px-4 w-full rounded-l-xl bg-primary-300 text-white"
      data-test-id="tab1Button"
      id="tab1Button"
      onClick={[Function]}
      type="button"
    >
      Tab 1
    </button>
    <button
      aria-label="Tab 2 label"
      className="py-1 px-4 w-full rounded-r-xl bg-primary-100 hover:bg-primary-300/80 hover:text-white text-black dark:text-white"
      data-test-id="tab2Button"
      id="tab2Button"
      onClick={[Function]}
      type="button"
    >
      Tab 2
    </button>
  </div>
  <div>
    <div
      className="h-full block"
    >
      Panel 1 
    </div>
  </div>
</div>
`;

exports[`Tabs when asked to render with children renders with children 1`] = `
<div
  className="flex flex-col"
>
  <div
    className="flex w-full rounded-xl"
  >
    <button
      aria-label="Tab 1 label"
      className="py-1 px-4 w-full rounded-l-xl bg-primary-300 text-white"
      data-test-id="tab1Button"
      id="tab1Button"
      onClick={[Function]}
      type="button"
    >
      Tab 1
    </button>
    <button
      aria-label="Tab 2 label"
      className="py-1 px-4 w-full rounded-r-xl bg-primary-100 hover:bg-primary-300/80 hover:text-white text-black dark:text-white"
      data-test-id="tab2Button"
      id="tab2Button"
      onClick={[Function]}
      type="button"
    >
      Tab 2
    </button>
  </div>
  <div>
    <div
      className="h-full block"
    >
      Panel 1 
    </div>
  </div>
</div>
`;

exports[`Tabs when asked to render without children renders without children 1`] = `
<div
  className="flex flex-col"
>
  <div
    className="flex w-full rounded-xl"
  />
  <div />
</div>
`;
